### PR TITLE
Manage enabled_cores_mask in snd_sof_dsp_core_power APIs

### PR DESCRIPTION
**Pull request by bardliao**
The goal is to avoid powering up/down enabled/disabled dsp cores.

Fixes: #2494

### CODE REVIEW

**Comment by bardliao** [comment]
Looks like my PR causes regression on ICL platform. I will check it tomorrow.

**Comment by bardliao** [comment]
SOFCI TEST

**Comment by bardliao** [comment]
> Looks like my PR causes regression on ICL platform. I will check it tomorrow.

Looks like it was a false failure. I can't reproduce the issue on a ICL HDA RVP and it is passed after re-triggering CI test.

File: sound/soc/sof/intel/hda-dsp.c [hunk 1]
```diff
@@ -624,7 +624,7 @@ static int hda_suspend(struct snd_sof_dev *sdev, bool runtime_suspend)
 #endif
 
 	/* power down DSP */
-	ret = hda_dsp_core_reset_power_down(sdev, chip->host_managed_cores_mask);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dsp.c:627]: @bardliao can we please add the ASOC: SOF: Intel: hda prefix in the commit title?

**Review by ranj063 — DISMISSED** [review]
LGTM. Thanks @bardliao

File: sound/soc/sof/intel/hda-dsp.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-dsp.c:627]: > @bardliao can we please add the ASOC: SOF: Intel: hda prefix in the commit title?

Done

File: sound/soc/sof/intel/hda-loader.c [hunk 2]
```diff
@@ -145,8 +145,10 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
 				       chip->ipc_ack_mask,
 				       chip->ipc_ack_mask);
 
-	/* step 5: power down corex */
-	ret = hda_dsp_core_power_down(sdev, chip->host_managed_cores_mask & ~(BIT(0)));
+	/* step 5: power down corex except init corex */
+	/* We should at least keep core 0 alive in case init_core_mask is not set */
+	ret = hda_dsp_core_power_down(sdev, chip->host_managed_cores_mask &
+				      ~(chip->init_core_mask | BIT(0)));
```
Comment (lyakh) [inline sound/soc/sof/intel/hda-loader.c]: I think you could use `SOF_DSP_PRIMARY_CORE` for that

**Review by lyakh — COMMENTED** [review]
we have the `SOF_DSP_PRIMARY_CORE` macro, but unfortunately it isn't used everywhere. In some places `0` is hard-coded. One of them is in this PR, we could replace it now.

File: sound/soc/sof/intel/hda-dsp.c [hunk 3]
```diff
@@ -624,7 +624,7 @@ static int hda_suspend(struct snd_sof_dev *sdev, bool runtime_suspend)
 #endif
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-dsp.c:624]: Commit message should explain a bit more. How about:
"-So we can manage enabled_cores_mask flag in the API."
"+To implement common logic in SOF core, core power up/down flows should use common SOF API and not directly use low-level platform specific helper functions."

File: sound/soc/sof/intel/hda-dsp.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda-dsp.c:627]: @ranj063 Although, the "hda:" prefix is not too helpful. I mean, we have patches with "hda:", "sdw:" and "byt:" prefixes, but platforms using sdw use these same core up/down helpers (and rest of HDA infra) as well. Not a big thing though...

File: sound/soc/sof/intel/hda-loader.c (hunk 2 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda-loader.c]: @bardliao what's with the term "corex", what does that mean? I thought it was a typo in existing code, but this extends its use, so I think it's worth explaining a bit more what is menat here

Another thing: it looks a bit odd that previous patches removes direct calls to core power up/down, but this one keeps it. Even if FW load is somewhat special flow, shouldn't we still use common API calls to do the actual power up/down, even here.

File: sound/soc/sof/ops.h [hunk 4]
```diff
@@ -68,19 +68,29 @@ static inline int snd_sof_dsp_reset(struct snd_sof_dev *sdev)
 static inline int snd_sof_dsp_core_power_up(struct snd_sof_dev *sdev,
```
Comment (kv2019i) [inline sound/soc/sof/ops.h:68]: @bardliao If API is in capital letters, DSP should be as well, otherwise it looks odd

The commit message is again a bit hard to follow. The main change is to move management of enable_core_mask to common API functions, so maybe start with that:

```
ASoC: SOF: update enabled_cores_mask only in core_power_up/down

Move management of enabled_cores_mask from topology load to common code 
in core_power_up/down to ensure it stays in sync with actual DSP core state.
```

If of course depends on whether we really do all power up/downs via snd_sof_dsp_core_power_up/down, or whether we still leave some special cases (like e.g. for FW boot). In that case, more explanation is needed.

File: sound/soc/sof/intel/hda-loader.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-loader.c]: > I think you could use `SOF_DSP_PRIMARY_CORE` for that

Yes, updated.

**Review by ranj063 — DISMISSED** [review]

**Review by kv2019i — CHANGES_REQUESTED** [review]
Hmm, I have a bit mixed feelings whether this goes far enough in cleanup. If we track core state in SOF common code, then I'd argue all calls to change core state (even during FW boot), should go via common code. Now we seem to still have a mix where some state is updated by common code, while some is done under the hood in vendor specific code, and the state data is in common sdev struct. Maybe I'm missing something w.r.t. FW boot and we have to keep it private, but please take a look whether this could be cleaned up even more.

**Comment by ranj063** [comment]
> Hmm, I have a bit mixed feelings whether this goes far enough in cleanup. If we track core state in SOF common code, then I'd argue all calls to change core state (even during FW boot), should go via common code. Now we seem to still have a mix where some state is updated by common code, while some is done under the hood in vendor specific code, and the state data is in common sdev struct. Maybe I'm missing something w.r.t. FW boot and we have to keep it private, but please take a look whether this could be cleaned up even more.

@kv2019i good point. I think we should go one step further to make the hda_dsp_power_down/up() functions static and ensure we dont use them at all and only the top-level

**Comment by bardliao** [comment]
> > Hmm, I have a bit mixed feelings whether this goes far enough in cleanup. If we track core state in SOF common code, then I'd argue all calls to change core state (even during FW boot), should go via common code. Now we seem to still have a mix where some state is updated by common code, while some is done under the hood in vendor specific code, and the state data is in common sdev struct. Maybe I'm missing something w.r.t. FW boot and we have to keep it private, but please take a look whether this could be cleaned up even more.
> 
> @kv2019i good point. I think we should go one step further to make the hda_dsp_power_down/up() functions static and ensure we dont use them at all and only the top-level

@kv2019i @ranj063 The naming of hda_dsp_core_power_up/down is confusing.  I thought they are the .core_power_up/down ops, but they are not.
The .core_power_up/down ops are actually
```
        .core_power_up = hda_dsp_enable_core,
        .core_power_down = hda_dsp_core_reset_power_down,
```
I was thinking updating core state in vendor specific code, but the core_power_up/down ops are not defined everywhere.  So sdev->enabled_cores_mask will be always 0 in the platforms which doesn't set core_power_up/down ops. Besides, snd_sof_dsp_core_power_up() is not called before FW booting in the existing code, but obviously, at least one dsp core will be powered up after FW boot. Therefore updating core state in snd_sof_dsp_core_power_up/down APIs is not sufficient.

File: sound/soc/sof/intel/hda-loader.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-loader.c]: > @bardliao what's with the term "corex", what does that mean? I thought it was a typo in existing code, but this extends its use, so I think it's worth explaining a bit more what is menat here

Sorry I don't know what "corex" means, either. But I see it appears multiple times in hda-loader.c. So I reuse the term.

> 
> Another thing: it looks a bit odd that previous patches removes direct calls to core power up/down, but this one keeps it. Even if FW load is somewhat special flow, shouldn't we still use common API calls to do the actual power up/down, even here.

I didn't change it. I would like to use common API calls here, too. But  hda_dsp_core_power_up/down is not equal to snd_sof_dsp_core_power_up/down. So it is complicated to use common API calls here. To be honesty, I don't know how to do it yet.

**Comment by bardliao** [comment]
@kv2019i I updated the commit messages. Does it look better now?

**Review by plbossart — DISMISSED** [review]
this looks fine, but the main comment I'd have is "are any Fixes tag required"? Do we want this to be applied to -stable? @bardliao @kv2019i what do you think?

**Comment by bardliao** [comment]
> this looks fine, but the main comment I'd have is "are any Fixes tag required"? Do we want this to be applied to -stable? @bardliao @kv2019i what do you think?

The issue exists for a long time. I don't mind to wait for another kernel merge cycle.

**Review by lyakh — DISMISSED** [review]

File: sound/soc/sof/intel/hda-dsp.c [hunk 5]
```diff
@@ -624,7 +624,7 @@ static int hda_suspend(struct snd_sof_dev *sdev, bool runtime_suspend)
...
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-dsp.c:625]: very minor: it's not common to end summary line to a period (".")

File: sound/soc/sof/intel/hda-loader.c [hunk 6]
```diff
@@ -145,8 +145,11 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
 				       chip->ipc_ack_mask,
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-loader.c:147]: very minor: it's not common to end summary line to a period (".")

File: sound/soc/sof/intel/hda-loader.c [hunk 7]
```diff
@@ -145,8 +145,11 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
...
 				       chip->ipc_ack_mask);
 
-	/* step 5: power down corex */
-	ret = hda_dsp_core_power_down(sdev, chip->host_managed_cores_mask & ~(BIT(0)));
+	/* step 5: power down corex except init corex */
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-loader.c]: @lgirdwood @plbossart It seems this "power-up corex" language has been there from the start and it's used multiple times so doesn't look accidental. What does the X mean and why we keep talking about "powering up corex". We seem to be using at as a plural of "core", which I don't understand. And above there's a new use "init corex" -- referring to single core. Could be English but seems confusing -- comments should reduce confusion, not add to it.

File: sound/soc/sof/ops.h [hunk 8]
```diff
@@ -70,7 +70,9 @@ static inline int snd_sof_dsp_core_power_up(struct snd_sof_dev *sdev,
 {
 	int ret = 0;
 
-	if (sof_ops(sdev)->core_power_up) {
+	/* Only power up the disabled cores */
```
Comment (kv2019i) [inline sound/soc/sof/ops.h]: Btw, I now realize "enabled_cores" is not the best variable name. On host side, if you disable a CPU core, it's put out of use (and will be powered off as well). I.e. generally enabling/disabling is not same as powering-up/down. In SOF, the two actions are the same, but reading the above comment, I find it confusing. Not really something this PR has added, so can be ignored here.

@ranj063 Any background info why you picked "enabled_cores" as the name?

**Review by kv2019i — CHANGES_REQUESTED** [review]
Sorry to nitpick, but there are still multiple minor confusing parts in commit messages and code comments that I think should be clarified. I especially find the "corex" language confusing. 

As for specially case handling for FW boot -- maybe that's ok, and probably makes sense to improve this incrementally. Any comments @ranj063 and @lyakh -- you've touched this code the most in recent times.

**Comment by bardliao** [comment]
Thanks @kv2019i  I updated the commit summary. Let's wait for comments for naming.

File: sound/soc/sof/intel/hda-loader.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c]: @kv2019i the X in the corex is just a substitue for the core IDs. FOr example on CNL, corex would be cores 2-4.

Comment (kv2019i) [reply sound/soc/sof/intel/hda-loader.c]: @ranj063 That still leaves we wondering what is X, and whether this adds value to understanding the code. It seems this is not as if I "open code" the new comment, it would be for CNL "step 5: power down cores 2-4 except init cores 2-4". So it doesn't make sense unless you just read "corex" as some set of cores. Oh well, I can live this, but it seems like something that might have made sense at start, but now with more and more people modifying the code, is no longer helpful.

I'd propose following "open-coding": "step 5: power down cores that are no longer needed"

File: sound/soc/sof/ops.h (hunk 8 above)
Comment (ranj063) [reply sound/soc/sof/ops.h]: @kv2019i I picked enabled_cores simply based on how we the core_power_up() op is set today:
	.core_power_up = hda_dsp_enable_core,

File: sound/soc/sof/intel/hda-loader.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c]: yes, sounds good. We should remove all ambiguities. In fact this step is a bit controversial already. For example on CNL+ platforms, you cannot really power off cores 2-4 ie writing 0 to the SPA bits for those cores has no effect.

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c]: maybe the host_managed_cores handles this. something to check

Comment (lyakh) [reply sound/soc/sof/intel/hda-loader.c]: FWIW "corex" confused me too. So, I understand another revision will be posted with those removed, also please remember to update the commit title accordingly

Comment (bardliao) [reply sound/soc/sof/intel/hda-loader.c]: @kv2019i @ranj063 @lyakh Updated.

**Review by lyakh — DISMISSED** [review]

File: sound/soc/sof/ops.h (hunk 8 above)
Comment (kv2019i) [reply sound/soc/sof/ops.h]: Ack @ranj063 It seems the PM core config IPC also has a field "enable_mask", so maybe it comes from there. 

@bardliao I think you should just drop the comment on line L73. In worst case it confuses (like me who mentally reads "disabled" to mean something else than powered off), and at best it doesn't add much new info.

File: sound/soc/sof/intel/hda-loader.c [hunk 9]
```diff
@@ -146,8 +146,10 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
 				       chip->ipc_ack_mask,
 				       chip->ipc_ack_mask);
 
-	/* step 5: power down corex */
-	ret = hda_dsp_core_power_down(sdev, chip->host_managed_cores_mask & ~(BIT(0)));
+	/* step 5: power down cores that are no longer needed */
+	ret = hda_dsp_core_power_down(sdev, chip->host_managed_cores_mask &
+				      ~(chip->init_core_mask |
+					BIT(SOF_DSP_PRIMARY_CORE)));
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-loader.c]: The git commit message is now in coflict with code. I agree with commit message that we can assume init_core_mask to contain the primary-core. With this assumption, the above can be simplified to:

ret = hda_dsp_core_power_down(sdev, chip->host_managed_cores_mask &
				      ~(chip->init_core_mask)));

File: sound/soc/sof/loader.c [hunk 10]
```diff
@@ -872,7 +872,8 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 	}
 
 	/* fw boot is complete. Update the active cores mask */
-	sdev->enabled_cores_mask = init_core_mask;
+	/* Primary core is active no matter init_core_mask is set or not */
+	sdev->enabled_cores_mask = init_core_mask | BIT(SOF_DSP_PRIMARY_CORE);
```
Comment (kv2019i) [inline sound/soc/sof/loader.c]: And same here.

File: sound/soc/sof/ops.h [hunk 11]
```diff
@@ -70,7 +70,9 @@ static inline int snd_sof_dsp_core_power_up(struct snd_sof_dev *sdev,
 {
```
Comment (kv2019i) [inline sound/soc/sof/ops.h:70]: Let's use this commit message to clarify the semantics. My proposal:

```
ASoC: SOF: Filter out unneeded core power up/downs

Exclude cores that are already powered on/off correctly. This allows to simplify
dsp_power_up/down() implementations.
```

Feel free to suggest to other wordings, but something to this effect would seem more understandable.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Sorry for dragging the review. I know @bardliao you are cleaning up existing code, but I keep having trouble following the logic we have to manage core state. I made a few more minor comments to the series.

File: sound/soc/sof/loader.c (hunk 10 above)
Comment (bardliao) [reply sound/soc/sof/loader.c]: Sorry, the commit message may be misleading. What I tried to say it that chip->init_core_mask is not set in all platforms.  In other words, chip->init_core_mask could be 0. That's why I tried to add the  SOF_DSP_PRIMARY_CORE bit to the  enabled_cores_mask. I agree that we don't need to`` | BIT(SOF_DSP_PRIMARY_CORE)`` in ``cl_dsp_init()`` since all platforms which set ``.run = hda_dsp_cl_boot_firmware,`` or ``.run = hda_dsp_cl_boot_firmware_iccmax,`` also set ``.init_core_mask``, too. But to let ``enabled_cores_mask`` match the real core power status, I think we should set  ``sdev->enabled_cores_mask = init_core_mask | BIT(SOF_DSP_PRIMARY_CORE)`` here.

**Comment by bardliao** [comment]
SOFCI TEST

**Comment by lyakh** [comment]
@bardliao have the same tests fail again?

**Comment by bardliao** [comment]
> @bardliao have the same tests fail again?

It failed on the same device (ICL RVP HDA) but different cases. I manually triggered a test plan on the the device to check the healthy  status of the device, but the testing is still on going. It has been 465 minute since the test plan is triggered.

**Comment by ranj063** [comment]
> > @bardliao have the same tests fail again?
> 
> It failed on the same device (ICL RVP HDA) but different cases. I manually triggered a test plan on the the device to check the healthy status of the device, but the testing is still on going. It has been 465 minute since the test plan is triggered.

@bardliao there have been new changes in ICL to include ICCMAX which requires core 3 to be powered up after FW boot. Can you please rebase your series on top of that?

**Comment by ranj063** [comment]
@bardliao take a look at this function: https://github.com/thesofproject/linux/blob/9f4458c17b30d22785ecb8db140017a39d18d2f5/sound/soc/sof/intel/hda-loader.c#L486

it does a stall after powering up the core 3. I wonder if this needs an unstall because of your changes. @fredoh9 FYI

**Comment by ranj063** [comment]
@bardliao take a look at this function: https://github.com/thesofproject/linux/blob/9f4458c17b30d22785ecb8db140017a39d18d2f5/sound/soc/sof/intel/hda-loader.c#L486

it does a stall after powering up the core 3. I wonder if this needs an unstall because of your changes. @fredoh9 FYI

**Comment by fredoh9** [comment]
I'm looking at it. I will test with this patches on ICL_RVP_HDA. Will update soon.

**Comment by fredoh9** [comment]
A quick update. Although unstalling core3 was the requirement, I added it and tested again. The result is same, same timeout error. Will do more debugging.

File: sound/soc/sof/loader.c (hunk 10 above)
Comment (ranj063) [reply sound/soc/sof/loader.c]: I think we shuldnt assume this but instead check if init_core_mask returned after FW boot is non-zero

**Comment by fredoh9** [comment]
Found the problem. Initially I suspect stall but it turned out misatch of enabled_cores_mask. For ICL ICCMAX, core 3 is powered up and stall at post_fw(). After core 3 power up, enabled_cores_mask became 8(=b1000) but this is overwritten by

```
        /* fw boot is complete. Update the active cores mask */
        /* Primary core is active no matter init_core_mask is set or not */
        sdev->enabled_cores_mask = init_core_mask | BIT(SOF_DSP_PRIMARY_CORE);
```

init_core_mask already have primary core enabled by hda_dsp_cl_boot_firmware,
https://github.com/thesofproject/linux/blob/cfa9a863e20a578dd2b2012880fe239cb6667f7c/sound/soc/sof/intel/hda-loader.c#L445

Not to lose enabled_cores_mask fromICL ICCMAX run op, I changed like below and this works fine. init_core_mask was 1 and updated enabled_cores_mask is 9(=b1001)

```
-       sdev->enabled_cores_mask = init_core_mask | BIT(SOF_DSP_PRIMARY_CORE);
+       sdev->enabled_cores_mask |= init_core_mask;
```

File: sound/soc/sof/loader.c (hunk 10 above)
Comment (fredoh9) [reply sound/soc/sof/loader.c]: we will lose enabled_cores_mask when enabled_core_mask is set before this. Actually ICL ICCMAX, we need to enable/stall core 3 before this. hda_dsp_cl_boot_firmware return primary core id. How about changing this to

```
-       sdev->enabled_cores_mask = init_core_mask | BIT(SOF_DSP_PRIMARY_CORE);
+       sdev->enabled_cores_mask |= init_core_mask;
```

**Review by fredoh9 — CHANGES_REQUESTED** [review]

**Comment by ranj063** [comment]
> Found the problem. Initially I suspect stall but it turned out misatch of enabled_cores_mask. For ICL ICCMAX, core 3 is powered up and stall at post_fw(). After core 3 power up, enabled_cores_mask became 8(=b1000) but this is overwritten by
> 
> ```
>         /* fw boot is complete. Update the active cores mask */
>         /* Primary core is active no matter init_core_mask is set or not */
>         sdev->enabled_cores_mask = init_core_mask | BIT(SOF_DSP_PRIMARY_CORE);
> ```
> 
> init_core_mask already have primary core enabled by hda_dsp_cl_boot_firmware,
> https://github.com/thesofproject/linux/blob/cfa9a863e20a578dd2b2012880fe239cb6667f7c/sound/soc/sof/intel/hda-loader.c#L445
> 
> Not to lose enabled_cores_mask fromICL ICCMAX run op, I changed like below and this works fine. init_core_mask was 1 and updated enabled_cores_mask is 9(=b1001)
> 
> ```
> -       sdev->enabled_cores_mask = init_core_mask | BIT(SOF_DSP_PRIMARY_CORE);
> +       sdev->enabled_cores_mask |= init_core_mask;
> ```

We also need to fix hda_dsp_cl_boot_firmware_iccmax() to return the init_core_mask upon success along with this change.

**Comment by ranj063** [comment]
> We also need to fix hda_dsp_cl_boot_firmware_iccmax() to return the init_core_mask upon success along with this change.

ok i was confused there for a bit. This is not needed but the right fix is to  move sdev->enabled_cores_mask = init_core_mask; before the call to snd_sof_dsp_post_fw_run() and we should check if this is not 0.

**Comment by ranj063** [comment]
> > We also need to fix hda_dsp_cl_boot_firmware_iccmax() to return the init_core_mask upon success along with this change.
> 
> ok i was confused there for a bit. This is not needed but the right fix is to move sdev->enabled_cores_mask = init_core_mask; before the call to snd_sof_dsp_post_fw_run() and we should check if this is not 0.

actually I take that back, we probably dont even need this anymore because the enabled_cores_mask should be updated autmatically with this PR

**Review by fredoh9 — COMMENTED** [review]
In 2nd commit message, I understand the point but use exact variable name, s/enable_cores_mask /enabled_cores_mask

**Comment by fredoh9** [comment]
> actually I take that back, we probably dont even need this anymore because the enabled_cores_mask should be updated autmatically with this PR

I totally agree. If we can remove manual update(or initializing with primary core) with PR, that will be great.
I added debug log in ICL to check this, but it looks enabled_cores_mask is not up to dated. This is a log from .post_fw_run.

```
[   13.717135] sof-audio-pci 0000:00:1f.3: hda_dsp_post_fw_run_icl: HPRO dsp core 3 power up, enabled_cores_mask 0
```

**Comment by bardliao** [comment]
> We also need to fix hda_dsp_cl_boot_firmware_iccmax() to return the init_core_mask upon success along with this change.

We [return chip_info->init_core_mask;](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-loader.c#L446)  in hda_dsp_cl_boot_firmware and [ret = hda_dsp_cl_boot_firmware(sdev);](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-loader.c#L316) is called by hda_dsp_cl_boot_firmware_iccmax. So I think it is good now.

**Comment by bardliao** [comment]
@fredoh9 @ranj063 The PR is updated, could you take a look?

**Comment by ranj063** [comment]
> @fredoh9 @ranj063 The PR is updated, could you take a look?

@bardliao what are the UP2 failures about? Are they related to your changes?

File: sound/soc/sof/intel/hda-loader.c [hunk 12]
```diff
@@ -93,7 +93,7 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
 	int i;
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-loader.c:93]: typo: orignial -> original

**Review by kv2019i — DISMISSED** [review]
Thanks @bardliao for the many updates. I think this looks good now. I found one typo, but no need to update the series for that. 

The APL-NOCODEC needs checking (not going to suspend), but if that's unrelated, ok to merge.

File: sound/soc/sof/intel/hda-loader.c [hunk 13]
```diff
@@ -148,8 +148,8 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
 				       chip->ipc_ack_mask);
 
 	/* step 5: power down cores that are no longer needed */
-	ret = hda_dsp_core_power_down(sdev, chip->host_managed_cores_mask &
-				      ~(chip->init_core_mask));
+	ret = snd_sof_dsp_core_power_down(sdev, chip->host_managed_cores_mask &
+					  ~(chip->init_core_mask));
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:152]: @bardliao now that we're doing this, we dont actually need to set enabled_cores_mask to init_cores in loader.c after FW boot is complete, right?

Comment (fredoh9) [reply sound/soc/sof/intel/hda-loader.c:152]: @bardliao I agreed with Ranjani. After this, enabled_cores_mask should have primary core set(chip->init_core_mask). So we don't need to initialize init_core_mask in snd_sof_run_firmware(). Other than this, all good to go.

Comment (bardliao) [reply sound/soc/sof/intel/hda-loader.c:152]: @ranj063 @fredoh9 Yes, that's why I removed all enabled_cores_mask setting outside snd_sof_dsp_core_power_down/up APIs.

File: sound/soc/sof/intel/hda-loader.c (hunk 12 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-loader.c:93]: > typo: orignial -> original

Updated

**Review by kv2019i — APPROVED** [review]

**Comment by ranj063** [comment]
@bardliao almost ready except for the one travis build failure

**Comment by bardliao** [comment]
> @bardliao almost ready except for the one travis build failure

The build failure should be fixed, but somehow the the Travis CI build is still in progress.

**Review by fredoh9 — APPROVED** [review]
LGTM

**Review by ranj063 — APPROVED** [review]

**Comment by bardliao** [comment]
@kv2019i Does it look good now?

**Review by kv2019i — APPROVED** [review]
